samcli/commands/publish/command.py

```diff
@@ -0,0 +1,131 @@
+"""CLI command for "publish" command."""
+
+import json
+
+import yaml
+import click
+import boto3
+from botocore.exceptions import ClientError
+from serverlessrepo import publish_application
+from serverlessrepo.publish import CREATE_APPLICATION
+from serverlessrepo.exceptions import ServerlessRepoError
+
+from samcli.cli.main import pass_context, common_options as cli_framework_options, aws_creds_options
+from samcli.commands._utils.options import template_common_option
+from samcli.commands._utils.template import get_template_data
+from samcli.commands.exceptions import UserException
+
+HELP_TEXT = """
+Use this command to publish a packaged AWS SAM template to
+the AWS Serverless Application Repository to share within your team,
+across your organization, or with the community at large.\n
+\b
+This command expects the template's Metadata section to contain an
+AWS::ServerlessRepo::Application section with application metadata
+for publishing. For more details on this metadata section, see
+https://docs.aws.amazon.com/serverlessrepo/latest/devguide/serverless-app-publishing-applications.html
+\b
+Examples
+--------
+To publish an application
+$ sam publish -t packaged.yaml --region <region>
+"""
+SHORT_HELP = "Publish a packaged AWS SAM template to the AWS Serverless Application Repository."
+SERVERLESSREPO_CONSOLE_URL = "https://console.aws.amazon.com/serverlessrepo/home?region={}#/published-applications/{}"
+
+
+@click.command("publish", help=HELP_TEXT, short_help=SHORT_HELP)
+@template_common_option
+@aws_creds_options
+@cli_framework_options
+@pass_context
+def cli(ctx, template):
+    # All logic must be implemented in the ``do_cli`` method. This helps with easy unit testing
+
+    do_cli(ctx, template)  # pragma: no cover
+
+
+def do_cli(ctx, template):
+    """Publish the application based on command line inputs."""
+    template_data = get_template_data(template)
+    try:
+        publish_output = publish_application(yaml.safe_dump(template_data))
```

Comment:
since you are using `get_template_data` you no longer need to do a `yaml.safe_dump` here. This is take care of by `get_template_data`

Reply:
publish_application expects a string as input, so I have to serialize it back to yaml string.

Reply:
I need to dump `template_data ` because `pubish_application` expects a YAML or JSON template as string. It doesn't accept template dictionary, which is why I didn't use `get_template_data` initially.

Reply:
The data is expected to be a python Dictionary here. So let's use json.dumps instead of yaml. We only need to use yaml if the data is yaml, which it is not.

Reply:
I see, I'll use json then.

Reply:
So that means we have no way to preserve the order of the template? Will the template in SAR display as YAML or JSON? We prefer YAML and ideally maintaining the same order the customer gave it to us in. I realize there are several places where we're currently parsing and scrambling the template order, but I'd like to eventually clean all of them up so when a customer builds/packages/publishes, the YAML template looks very close to what it looks like when they edit it. Scrambling the order makes it very hard to read.

Reply:
@jlhood That is fair. We should fix this across then instead on one offing, especially when we have one place that does the reading. 

Currently `publish_application` takes a string instead of a dict. If SAR wants to truly control this, I think the `publish_application` should take an OrderedDict and do the serialization instead of relying on the caller to do this action (keeping the template the same).